sdk/metric/sdk.go

```diff
@@ -358,25 +557,68 @@ func (m *SDK) Collect(ctx context.Context) int {
 		// TODO: Reconsider this logic.
 		if inuse.refMapped.inUse() || atomic.LoadInt64(&inuse.modified) != 0 {
 			atomic.StoreInt64(&inuse.modified, 0)
-			checkpointed += m.checkpoint(ctx, inuse)
+			checkpointed += m.checkpointRecord(ctx, inuse)
 		}
 
 		// Always continue to iterate over the entire map.
 		return true
 	})
 
-	m.currentEpoch++
 	return checkpointed
 }
 
-func (m *SDK) checkpoint(ctx context.Context, r *record) int {
-	if r.recorder == nil {
+func (m *SDK) collectObservers() int {
+	checkpointed := 0
+
+	m.observers.Range(func(key, value interface{}) bool {
+		obs := key.(*observer)
+		cb := obs.getCallback()
+		result := observerResult{
+			observer: obs,
+		}
+		cb(result)
+		checkpointed += m.checkpointObserver(obs)
+		return true
+	})
+
+	return checkpointed
+}
+
+func (m *SDK) checkpointRecord(ctx context.Context, r *record) int {
+	return m.checkpoint(ctx, r.descriptor, r.recorder, r.labels)
+}
+
+func (m *SDK) checkpointObserver(obs *observer) int {
+	if len(obs.recorders) == 0 {
 		return 0
 	}
-	r.recorder.Checkpoint(ctx, r.descriptor)
-	labels := export.NewLabels(r.labels.sorted, r.labels.encoded, m.labelEncoder)
-	err := m.batcher.Process(ctx, export.NewRecord(r.descriptor, labels, r.recorder))
+	checkpointed := 0
+	ctx := context.Background()
```

Comment:
No clue what I was thinking when I wrote it - too much time has passed since then. :P Probably got carried away with thinking about observers differing from other instruments, that observer callbacks lack the context other instruments have at the time of recording the measurement.

But you are right, will fix it.